src/jit/codegenarm64.cpp

```diff
@@ -6476,287 +6483,388 @@ void CodeGen::genPutArgStk(GenTreePtr treeNode)
+                    // We will treat this case the same as above 
+                    // (i.e if we just had this GT_LCL_VAR directly as the source)
+                    // so update 'source' to point this GT_LCL_VAR_ADDR node
+                    // and continue to the codegen for the LCL_VAR node below
+                    //
+                    varNode  = addrNode->AsLclVarCommon();
+                    addrNode = nullptr;
+                }
             }
-            else  // We have a GT_OBJ with an address expression
+
+            // Either varNode or addrNOde must have been setup above,
+            // the xor ensures that only one of the two is setup, not both
+            assert((varNode != nullptr) ^ (addrNode != nullptr));
+
+            BYTE     gcPtrs[MAX_MULTIREG_ARGS] = {TYPE_GC_NONE, TYPE_GC_NONE, TYPE_GC_NONE, TYPE_GC_NONE };
+
+            BYTE*    structGcLayout = &gcPtrs[0];  // The GC layout for the struct
+            unsigned gcPtrCount;                   // The count of GC pointers in the struct
+            int      structSize;
+            bool     isHfa;
+
+            // Setup the structSize, isHFa, and gcPtrCount
+            if (varNode != nullptr)
             {
+                varNumInp = varNode->gtLclNum;
+                assert(varNumInp < compiler->lvaCount);
+                LclVarDsc* varDsc = &compiler->lvaTable[varNumInp];
+
+                assert(varDsc->lvType = TYP_STRUCT);
+                assert(varDsc->lvOnFrame);       // This struct also must live in the stack frame
+
+                structSize = varDsc->lvSize();   // This yields the roundUp size, but that is fine
+                                                 // as that is how much stack is allocated for this LclVar
+                isHfa = varDsc->lvIsHfa();
+                gcPtrCount = varDsc->lvStructGcCount;
+                structGcLayout = varDsc->lvGcLayout;               
+            }
+            else  // addrNode is used
+            {
+                assert(addrNode != nullptr);
+
                 // Generate code to load the address that we need into a register
                 genConsumeAddress(addrNode);
+                addrReg = addrNode->gtRegNum;
 
-                regNumber addrReg    = addrNode->gtRegNum;
-                var_types targetType = objNode->TypeGet();
-
-                noway_assert(varTypeIsStruct(targetType)); 
+                CORINFO_CLASS_HANDLE objClass = source->gtObj.gtClass;
+                var_types  hfaType;
 
-                CORINFO_CLASS_HANDLE objClass = objNode->gtObj.gtClass;
-                int structSize = compiler->info.compCompHnd->getClassSize(objClass);
-
-                assert(structSize <= 2*TARGET_POINTER_SIZE);
+                structSize = compiler->info.compCompHnd->getClassSize(objClass);
+                hfaType    = compiler->GetHfaType(objClass);   // set to float or double if it is an HFA, otherwise TYP_UNDEF
+                isHfa      = (hfaType != TYP_UNDEF);
```

Comment:
hfaType is not serving much use and all we need to know whether the struct being passed is hfa.  Hence the following would suffice.

isHfa = compiler->IsHfa(objClass).